### DISCUSSION TITLE
How can I enable rate limiting based on http header values?

### DISCUSSION THREAD

**Discussion by shahamit** [Kong/kubernetes-ingress-controller#1822] [Q&A] [answered]
We want to apply different rate limits based on the value of the custom header in the api request.

From the rate-limiting plugin config I see the property to enable header based rate limiting but there is no provision to configure different rate limits for different values of the header parameter. Could someone please let me know if this is possible and guide me further on the steps?

Thanks.

**Comment by shaneutt** [comment, marked as answer]
There are a few ways you could do this including:

1) Using the `headers` criteria of a [Kong Route](https://docs.konghq.com/gateway-oss/2.4.x/admin-api/#route-object)
2) a [Custom Kong Plugin](https://docs.konghq.com/kubernetes-ingress-controller/1.3.x/guides/setting-up-custom-plugins/#main)

Unfortunately `#1` above is blocked as `headers` was added recently in Kong `v2.4` but is not available in the Ingress controller yet:

https://github.com/Kong/kubernetes-ingress-controller/issues/1835

So for the moment a custom plugin with custom code would be needed until that issue is resolved.

**Reply by shahamit** [reply]
Thanks @shaneutt for responding back.

1. I have been through Kong Route object documentation. I didn't follow how would it help in applying different rate limits for different header parameter values. For e.g. Assuming we have a 'tenant-id' as one of the header parameters. I want to configure that 'tenant-id' 123 should allow rate limits as 5 api calls per second while 'tenant-id' value as 789 should allow rate limits as 10 api calls per second. How do I do this with route object?
2. In a custom plugin implementation how can I configure standard plugins? Considering the above example, can I programmatically instantiate and configure 2 rate-limiting plugin instances for 2 values of the 'tenant-id' parameter?

 Thanks.

**Reply by shaneutt** [reply]
> I have been through Kong Route object documentation. I didn't follow how would it help in applying different rate limits for different header parameter values. For e.g. Assuming we have a 'tenant-id' as one of the header parameters. I want to configure that 'tenant-id' 123 should allow rate limits as 5 api calls per second while 'tenant-id' value as 789 should allow rate limits as 10 api calls per second. How do I do this with route object?

The `headers` criteria would allow you to split routes based on specific headers, thus you would be able to add rate limits specific to those routes as I was expecting was the original request. Does that make sense, have I missed something? Let me know if that is helpful.

> In a custom plugin implementation how can I configure standard plugins? Considering the above example, can I programmatically instantiate and configure 2 rate-limiting plugin instances for 2 values of the 'tenant-id' parameter?

Yes you could make two separate plugins each configured for a specific value of `tenant-id` and then applying the rate limit. It should be possible with a custom plugin to do this as a single plugin as well.

**Reply by shahamit** [reply]
@shaneutt - What lua code do I write in my custom plugin to call the rate-limiting plugin that applies the rate limits? It would be great if you could share some code sample or point me to an existing one. 
Thanks.

**Reply by shaneutt** [reply]
I don't have any specific examples, but I do have some links that may be helpful to you:

- https://docs.konghq.com/gateway-oss/2.6.x/plugin-development/
- https://github.com/Kong/kong/tree/2.6.0/kong/plugins/rate-limiting

Additionally if you're using Kong Enterprise custom plugin development is offered for Enterprise customers you can reach out to your account manager directly and they can work with you on that.
